onnx/shape_inference/implementation.cc

```diff
@@ -294,33 +294,41 @@ class ShapeInferenceImplBase {
     if (checker::check_is_experimental_op(n)) {
       has_experimental_op = true;
     } else if (n.op_type() == "Constant" && n.output().size() == 1) {
+      const std::string& output_name = n.output(0);
       for (const auto& attr : n.attribute()) {
         if (attr.name() == "value") {
           if (attr.type() == AttributeProto::TENSOR && attr.has_t()) {
-            input_data_by_name[n.output(0)] = &attr.t();
+            if (reuse_constant_tensors) {
+              input_data_by_name[output_name] = &attr.t();
+            } else {
+              input_data_by_name_holder[output_name] = attr.t();
+              input_data_by_name[output_name] = &input_data_by_name_holder[output_name];              
+            }
           } else if (attr.type() == AttributeProto::SPARSE_TENSOR && attr.has_sparse_tensor()) {
-            input_sparse_data_by_name[n.output(0)] = &attr.sparse_tensor();
+            if (reuse_constant_tensors) {
+              input_sparse_data_by_name[output_name] = &attr.sparse_tensor();
+            }
           }
         } else {
           switch (attr.type()) {
             case AttributeProto::INTS: {
               std::vector<int64_t> ints{attr.ints().begin(), attr.ints().end()};
-              addTemporaryConstant(n.output(0), ints);
+              addTemporaryConstant(output_name, ints);
               break;
             }
             case AttributeProto::INT: {
               std::vector<int64_t> ints({attr.i()});
-              addTemporaryConstant(n.output(0), ints);
+              addTemporaryConstant(output_name, ints);
               break;
             }
             case AttributeProto::FLOATS: {
               std::vector<float> floats{attr.floats().begin(), attr.floats().end()};
-              addTemporaryConstant(n.output(0), floats);
+              addTemporaryConstant(output_name, floats);
               break;
             }
             case AttributeProto::FLOAT: {
               std::vector<float> floats({attr.f()});
-              addTemporaryConstant(n.output(0), floats);
+              addTemporaryConstant(output_name, floats);
```

Comment:
Just out of curiosity: I saw you have handled the case of tensor and sparse_tensor with `reuse_constant_tensors` above(L301-310). Why don't we need to handle similar things in this else condition?

Reply:
They already do it. The main fix is to create a temporary-constant (of the TensorProto, which is stored in `input_data_by_name_holder`), which already happens in `addTemporaryConstant`.

This is to fix wrong behavior. The problem is that we are storing the address of `attr.t()`, but this is stack-allocated temporary (see [this line](https://github.com/onnx/onnx/blob/main/onnx/shape_inference/implementation.cc#L552)) which will disappear once the Constant node is processed.

Reply:
The if-condition is for performance. We don't need the fix when doing inference for a Graph, only when we do inference for a Function, which is when we create the temporary NodeProto on stack.